ws/config.py

```diff
@@ -1,180 +1,232 @@
+    if string in true_values:
         return True
-    elif string in ["0", "false", "no"]:
+    elif string in false_values:
         return False
     else:
-        print("value '%s' cannot be converted to boolean" % value)
-        raise configargparse.ArgumentTypeError("value '%s' cannot be converted to boolean" % value)
-
-# any path, the dirname part must exist (e.g. path to a file that will be created in the future)
-def argtype_dirname_must_exist(string):
-    string = os.path.abspath(os.path.expanduser(string))
-    dirname, _ = os.path.split(string)
-    if not os.path.isdir(dirname):
-        raise configargparse.ArgumentTypeError("directory '%s' does not exist" % dirname)
-    return string
-
-# path to existing directory
-def argtype_existing_dir(string):
-    string = os.path.abspath(os.path.expanduser(string))
-    if not os.path.isdir(string):
-        raise configargparse.ArgumentTypeError("directory '%s' does not exist" % string)
-    return string
-
-# list of comma-separated items from a fixed set
-def argtype_comma_list_choices(choices):
-    choices = set(choices)
-
-    def wrapped(string):
-        items = [item.strip() for item in string.split(",")]
-        for item in items:
-            if item not in choices:
-                raise configargparse.ArgumentTypeError("unknown item: '{}' (available choices: {})".format(item, choices))
-        return items
-
-    return wrapped
-
-def getArgParser(subname=None, *args, **kwargs):
-    if subname is None:
-        _, _script = os.path.split(sys.argv[0])
-        subname, _ = os.path.splitext(_script)
-
-    # create config file parser
-    cfp = ConfigFileParser("site", subname)
-    kwargs["config_file_parser"] = cfp
-    kwargs["ignore_unknown_config_file_keys"] = True
-
-    # set brief usage parameter by default
-    kwargs.setdefault("usage", "%(prog)s [options]")
-    # TODO: the supplied description is merged with the one from configargparse, which looks very ugly, regardless of the used formatter
-    kwargs.setdefault("formatter_class", configargparse.RawDescriptionHelpFormatter)
-
-    ap = configargparse.ArgParser(*args, **kwargs)
-
-    # add config file and site arguments
-    ap.add_argument("-c", "--config", metavar="PATH", is_config_file_arg=True,
-            help="path to config file (default: %(default)s)")
-    ap.add_argument("--site",
-            help="sets the top-level section to be read from config files (default: %(default)s)")
+        raise argparse.ArgumentTypeError(
+            "cannot convert '{}'to boolean".format(string))
+
+
+def argtype_configfile(string):
+    """Convert `--config` argument to an absolute file path."""
+    dirname = os.path.dirname(string)
+    name, ext = os.path.splitext(os.path.basename(string))
+
+    # Configuration name was specified.
+    if not dirname and not ext:
+        config_dir = os.getenv('XDG_CONFIG_HOME',
+                               os.path.expanduser(CONFIG_DIR))
```

Comment:
`os.path.expanduser` should be used when `CONFIG_DIR` is defined.

Reply:
`CONFIG_DIR` is defined in the start of the file:
```python
CONFIG_DIR = "~/.config/"
```

Reply:
I meant that it should be defined as `CONFIG_DIR = os.path.expanduser("~/.config/")` so that you don't have to call `expanduser` later.

Reply:
I thought it is C-like mnemonic macros for the string and number literals. They [are used ](https://github.com/python/cpython/blob/3.8/Lib/argparse.py#L94)for convenience - if you want to change, for example, default config dir, you should do it only in one place (and no matter how much times it is used across the module). Function usage for global macro seem wrong to me.

Reply:
They are just global variables, there is nothing resembling C macros (except maybe the uppercase style). Using `os.path.expanduser` at the definition place avoids the need to use it everywhere else, so it seems very convenient to me. I don't see any reason against doing this while you can.

Reply:
I understand the difference between Python global variables and C macros, but I meant that it is used for defining the literals (and the uppercase style hints on it). We can use functions in it, but it is wrong - theoretically there is can be another usage for `"~/.config"` literal besides `os.path.expanduser` (ok, I can hardly come up with a suitable example, but _in theory_ it should work like this).

If you insist, then I will add this function, but I must warn you that this change will lead to a change in the entropy of the universe and unpredictable consequences that will be on your conscience. :-P

Reply:
I really think that at the point where you use `CONFIG_DIR`, you shouldn't be bothered with considering whether the variable contains `~` and thus you should use `os.path.expanduser`, just because you made the assumption that global variables are string literals. If we ever changed `CONFIG_DIR` such that it does not contain `~`, there would be useless calls to `os.path.expanduser` scattered all over the place, unless somebody remembered this discussion and removed them too - but this is against the principle that changes should be done in one place only. OK, this argument is more applicable to `CACHE_DIR` rather than `CONFIG_DIR`, but still...

Reply:
> unless somebody remembered this discussion and removed them too

Anyway, I do not think that to delete `os.path.expanduser()` calls will become a good idea one day. Actually, we can take path parameters in three manners:

- tilde started path
- relative path
- absolute path

Construct `os.path.abspath(os.path.expanduser(path))` is the only reliable way to convert an argument of any of the above to the valid absolute path (although in case of absolute path passed no conversion is required at all). Moreover, there is a downside: if you move `expanduser` from code to `CONFIG_DIR` assignment, the construct `os.path.abspath(CONFIG_DIR)` could lead to ambiguity. The reader will wonder every time has config path already been tilde expanded or not. "Explicit is better than implicit" (c) `import this`

In general, I slept well and I don't want to argue anymore. I'll add the call to the variable, although I will remain unconvinced. :/

Reply:
Mm, in fact, the text does not contain such `os.path.abspath(os.path.expanduser())` constructions with `CONFIG_DIR`, so I will add a little explanation of what I mean:

```python
# configuration name was specified
if not dirname and not ext:
config_dir = os.getenv("XDG_CONFIG_HOME", CONFIG_DIR)      # expanduser already moved
path = os.path.join(config_dir,
                        "{}/{}.conf".format(PROJECT_NAME, string))
# relative or absolute path was specified
else:
     if ext != ".conf":
         raise argparse.ArgumentTypeError(
             "config filename must end with '.conf' suffix")
         path = os.path.abspath(os.path.expanduser(string))              # expanduser can't be moved
```

On the last line we know for sure, that `string` will be tilde expanded. In the line of `config_dir = ` there is no such confidence - we must _constantly remember_ that `CONFIG_DIR` has already been tilde expanded. I cannot call it convenient.

Reply:
Converting every path to absolute is not necessary - the scripts never call `os.chdir` so relative paths will work just fine. If the user passes a relative path and you give him back an absolute path in some error message, the user will be confused.

Reply:
For some reason I thought that `os.path.isdir()` function and the like takes only an absolute path as an argument. Yeah, it was a mistake.

> If the user passes a relative path and you give him back an absolute path in some error message, the user will be confused.

A reasonable remark. Now, in view of the above, it makes sense to remove `os.path.abspath()` calls where it is not strictly necessary.